core/domain/story_services.py

```diff
@@ -613,3 +617,247 @@ def record_completed_node_in_story_context(user_id, story_id, node_id):
     if node_id not in progress_model.completed_node_ids:
         progress_model.completed_node_ids.append(node_id)
         progress_model.put()
+
+
+def get_story_rights_from_model(story_rights_model):
+    """Constructs a StoryRights object from the given story rights model.
+
+    Args:
+        story_rights_model: StoryRightsModel. Story rights from the
+            datastore.
+
+    Returns:
+        StoryRights. The rights object created from the model.
+    """
+
+    return story_domain.StoryRights(
+        story_rights_model.id,
+        story_rights_model.manager_ids,
+        story_rights_model.story_is_published
+    )
+
+
+def publish_story(story_id, committer_id):
+    """Marks the given story as published.
+
+    Args:
+        story_id: str. The id of the given story.
+        committer_id: str. ID of the committer.
+
+    Raises:
+        Exception. The given story does not exist.
+        Exception. The story is already published.
+        Exception. The user does not have enough rights to publish the story.
+    """
+    def _is_node_valid_for_publishing(node):
+        if not node.exploration_id:
+            raise Exception(
+                'Story node does not contain an exploration id.')
+        if not exp_services.get_exploration_by_id(
```

Comment:
These get_exploration_by_id() and get_exploration_rights() calls are being done one at a time, which can be quite expensive. Can we batch them and do a get_multi() instead to retrieve the models from the datastore?

Reply:
Currently, there's only one API call done here i.e. for `initial_node_id`

Reply:
I don't think that's true, unless I'm misunderstanding something. This runs in a "for node in story.story_contents.nodes" loop (see below), so why do you say that there is only one API call?

Reply:
Yes, the loop runs for all the nodes but the function `_is_node_valid_for_publishing` is called only if the condition `node.id == story.story_contents.initial_node_id` is matched and it's correct only once.

Reply:
Ah, I see, thanks for explaining. I think it's too far away from this logic, though, and in the future it may be modified to check more than one node.

Why not replace with a get_multi() instead? That way, if the logic criterion changes, we will still do the fetch efficiently -- I don't see a downside.